src/components/app/App.jsx

```diff
@@ -1,51 +1,31 @@
 import React from 'react';
-import { useState } from 'react';
+import { useReducer } from 'react';
+import reducer, { initialState }  from '../reducers/colorReducer';
 
 
-const useRecord = (init) => {
-  const [before, setBefore] = useState([]);
-  const [current, setCurrent] = useState(init);
-  const [after, setAfter] = useState([]);
+export default function App() {
+  const [state, dispatch] = useReducer(reducer, initialState);
 
-  const undo = () => {
-    setAfter(after => [current, ...after]);
-    setCurrent(before[before.length - 1]);
-    setBefore(before => before.slice(0, -1));
-  };
-
-  const redo = () => {
-    setBefore(before => [...before, current]);
-    setCurrent(after[0]);
-    setAfter(after => after.slice(1));
-  };
-
-  const record = val => {
-    setBefore(before => [...before, current]);
-    setCurrent(val);
-  };
-
-  return {
-    undo,
-    record,
-    redo,
-    current,
-  };
-};
-
-function App() {
-  const { current, undo, redo, record } = useRecord('#FF0000');
+  
+  const record = (value) => dispatch({ type: 'RECORD', 
+    payload: value });
+  
 
   return (
     <>
-      <button data-testid="undo" onClick={undo}>undo</button>
-      <button data-testid="redo" onClick={redo}>redo</button>
-      <input data-testid="current" type="color" value={current} 
+      <button data-testid="undo" onClick={() => dispatch({
+        type: 'UNDO'
+      })}>undo</button>
+      <button data-testid="redo" onClick={() => dispatch({
+        type: 'REDO'
+      })}>redo</button>
+      <input data-testid="current" type="color" value={state.current} 
```

Comment:
Needs to not let the user advance passed the array in state for undo and redo there are two ways of doing this 
1.  disable the buttons by  
`<button data-testid="undo" disabled={!before.length} onClick={() => dispatch({ type: 'UNDO'  })}>undo</button>` 
this will disable the button when there is nothing in the array
1. Or in the reducer 
` case 'REDO':
      return {
        ...state,
        before: [...state.before, ...state.current],
        current: state.after[0] || ['#FFFF00'],
        after: state.after.slice(1),
      };`
If state is undefined it will set it to a default color